mixins/property-required/test/property-required-mixin.test.js

```diff
@@ -0,0 +1,219 @@
+
+	it('should throw if no property is defined', async() => {
+		const elem = await fixture(`<${tagValid}></${tagValid}>`);
+		expect(() => elem.addRequiredProperty('missing'))
+			.to.throw(createUndefinedPropertyMessage(tagValid, 'missing'));
+	});
+
+	it('should throw if type is non-String', async() => {
+		const tag = defineCE(
+			class extends PropertyRequiredMixin(LitElement) {
+				static properties = {
+					attr: { type: Boolean }
+				};
+			}
+		);
+		const elem = await fixture(`<${tag}></${tag}>`);
+		expect(() => elem.addRequiredProperty('attr'))
+			.to.throw(createInvalidPropertyTypeMessage(tag, 'attr'));
+	});
+
+	it('should use custom message', async() => {
+		const tag = defineCE(
+			class extends PropertyRequiredMixin(LitElement) {
+				static properties = {
+					attr: { type: String }
+				};
+				constructor() {
+					super();
+					this.addRequiredProperty('attr', {
+						message: () => 'custom message'
+					});
+				}
+			}
+		);
+		const elem = await fixture(`<${tag}></${tag}>`);
+		expect(() => elem.flushRequiredPropertyErrors())
+			.to.throw(TypeError, 'custom message');
+	});
+
+	it('should use custom validator', async() => {
+		const tag = defineCE(
+			class extends PropertyRequiredMixin(LitElement) {
+				static properties = {
+					attr: { type: String }
+				};
+				constructor() {
+					super();
+					this.addRequiredProperty('attr', {
+						validator: () => this.attr === 'valid'
+					});
+				}
+			}
+		);
+		const elem = await fixture(`<${tag} attr="oh no"></${tag}>`);
+		expect(() => elem.flushRequiredPropertyErrors())
+			.to.throw(TypeError, createDefaultMessage(tag, 'attr'));
+	});
+
+	it('should pass hasValue to custom validator', async() => {
+		const tag = defineCE(
+			class extends PropertyRequiredMixin(LitElement) {
+				static properties = {
+					attr: { type: String }
+				};
+				constructor() {
+					super();
+					this.addRequiredProperty('attr', {
+						validator: hasValue => hasValue
+					});
+				}
+			}
+		);
+		const elem = await fixture(`<${tag}></${tag}>`);
+		expect(() => elem.flushRequiredPropertyErrors())
+			.to.throw(TypeError, createDefaultMessage(tag, 'attr'));
+	});
+
+	it('should validate when dependent property changes', async() => {
+		const tag = defineCE(
+			class extends PropertyRequiredMixin(LitElement) {
+				static properties = {
+					attr1: { type: String },
+					attr2: { type: String }
+				};
+				constructor() {
+					super();
+					this.addRequiredProperty('attr1', {
+						dependentProps: ['attr2'],
+						validator: () => this.attr2 === 'valid'
+					});
+				}
+			}
+		);
+		const elem = await fixture(`<${tag} attr1="value" attr2="valid"></${tag}>`);
+		expect(() => elem.flushRequiredPropertyErrors()).to.not.throw();
+		elem.removeAttribute('attr2');
+		await elem.updateComplete;
+		expect(() => elem.flushRequiredPropertyErrors())
+			.to.throw(TypeError, createDefaultMessage(tag, 'attr1'));
+	});
+
+	it('should work in a subclass/mixin', async() => {
```

Comment:
This was a crazy case. When mixins or inheritence are present, the static `properties` need to be collected by traversing the inheritence tree. Loosly inspired by @eKoopmans' [provider stuff](https://github.com/Brightspace/unified-content-experience/blob/main/src/utilities/provider-framework/provider-mixin/provider-mixin.js#L41).